src/RageFileBasic.cpp

```diff
@@ -84,145 +84,145 @@ int RageFileObj::Seek( int iOffset )
-	if( iPos != -1 )
-	        m_iFilePos = iPos;
+	int iPos = SeekInternal(iOffset);
+	if (iPos != -1)
+		m_iFilePos = iPos;
 	return iPos;
 }
 
-int RageFileObj::Seek( int offset, int whence )
+int RageFileObj::Seek(int offset, int whence)
 {
-	switch( whence )
+	int iOffset = static_cast<int>(offset);
+	switch (whence)
 	{
 	case SEEK_CUR:
-		return Seek( Tell() + offset );
+		return Seek(Tell() + iOffset);
 	case SEEK_END:
-		offset += GetFileSize();
+		iOffset += GetFileSize();
 	}
-	return Seek( (int) offset );
+	return Seek(iOffset);
 }
 
-int RageFileObj::Read( void *pBuffer, std::size_t iBytes )
+int RageFileObj::Read(void* pBuffer, std::size_t iBytes)
 {
-	int iRet = 0;
+	char* pDest = static_cast<char*>(pBuffer);
+	std::size_t iTotalRead = 0;
 
-	while( !m_bEOF && iBytes > 0 )
+	while (!m_bEOF && iBytes > 0)
 	{
-		if( m_pReadBuffer != nullptr && m_iReadBufAvail )
+		if (m_pReadBuffer != nullptr && m_iReadBufAvail)
 		{
 			/* Copy data out of the buffer first. */
-			int iFromBuffer = std::min( (int) iBytes, m_iReadBufAvail );
-			memcpy( pBuffer, m_pReadBuf, iFromBuffer );
-			if( m_bCRC32Enabled )
-				CRC32( m_iCRC32, pBuffer, iFromBuffer );
+			int iFromBuffer = std::min(static_cast<int>(iBytes), m_iReadBufAvail);
+			memcpy(pDest, m_pReadBuf, iFromBuffer);
+			if (m_bCRC32Enabled)
+				CRC32(m_iCRC32, pDest, iFromBuffer);
 
-			iRet += iFromBuffer;
 			m_iFilePos += iFromBuffer;
 			iBytes -= iFromBuffer;
 			m_iReadBufAvail -= iFromBuffer;
 			m_pReadBuf += iFromBuffer;
 
-			pBuffer = (char *) pBuffer + iFromBuffer;
+			pDest += iFromBuffer;
 		}
 
-		if( !iBytes )
+		if (!iBytes)
 			break;
 
-		ASSERT( m_iReadBufAvail == 0 );
+		ASSERT(m_iReadBufAvail == 0);
 
 		/* If buffering is disabled, or the block is bigger than the buffer,
-		 * read the remainder of the data directly into the desteination buffer. */
-		if( m_pReadBuffer == nullptr || iBytes >= BSIZE )
+		 * read the remainder of the data directly into the destination buffer. */
+		if (m_pReadBuffer == nullptr || iBytes >= BSIZE)
 		{
 			/* We have a lot more to read, so don't waste time copying it into the
 			 * buffer. */
-			int iFromFile = this->ReadInternal( pBuffer, iBytes );
-			if( iFromFile == -1 )
+			int iFromFile = this->ReadInternal(pDest, iBytes);
+			if (iFromFile == -1)
 				return -1;
 
-			if( iFromFile == 0 )
+			if (iFromFile == 0)
 				m_bEOF = true;
 
-			if( m_bCRC32Enabled )
-				CRC32( m_iCRC32, pBuffer, iFromFile );
-			iRet += iFromFile;
+			if (m_bCRC32Enabled)
+				CRC32(m_iCRC32, pDest, iFromFile);
+
 			m_iFilePos += iFromFile;
-			return iRet;
+			iTotalRead += iFromFile;
+			return iTotalRead;
 		}
 
 		/* If buffering is enabled, and we need more data, fill the buffer. */
 		m_pReadBuf = m_pReadBuffer;
 		int iGot = FillReadBuf();
-		if( iGot == -1 )
+		if (iGot == -1)
 			return iGot;
-		if( iGot == 0 )
+		if (iGot == 0)
 			m_bEOF = true;
 	}
 
-	return iRet;
+	iTotalRead += (pDest - static_cast<char*>(pBuffer));
```

Comment:
We're subtracting the memory addresses? I don't think we should be doing that.

Reply:
The idea here is that casting pBuffer to a char* gives the number of bytes in the buffer. It is not subtracting the elements within the buffer.